[WIP] webhooks: Update NewRelic webhook for new format

maxrzaw opened the pull request:
<!-- What's this PR for?  (Just a link to an issue is fine.) -->
Fixes: #16393.

**Testing plan:** <!-- How have you tested? -->
Updated tests to match the updated format specified in the [New Relic Docs](https://docs.newrelic.com/docs/alerts-applied-intelligence/new-relic-alerts/alert-notifications/customize-your-webhook-payload#variables). 
Passes all test and has 100% coverage.
Made an account on New Relic to get some real testing data and tested using curl.

**GIFs or screenshots:** <!-- If a UI change.  See:
  https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
  -->
<img width="600" alt="NewRelicIntegrationExample" src="https://user-images.githubusercontent.com/32819501/99039869-226e4800-2556-11eb-9cdf-dc87e5892017.png">

<!-- Also be sure to make clear, coherent commits:
  https://zulip.readthedocs.io/en/latest/contributing/version-control.html
  -->

**Update 11/14/20:**
Testing plan is unchanged
**GIFs or screenshots:** <!-- If a UI change.  See:
  https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
  -->
Output format has been changed due to feedback [in zulip chat](https://chat.zulip.org/#narrow/stream/31-production-help/topic/New.20Relic)
<img width="600" alt="Screen Shot 2020-11-14 at 4 26 33 PM" src="https://user-images.githubusercontent.com/32819501/99157420-a7607b00-2696-11eb-9fed-7021b32a05b7.png">
Individual issues will get their own topic:
<img width="253" alt="Screen Shot 2020-11-14 at 4 28 57 PM" src="https://user-images.githubusercontent.com/32819501/99157485-2d7cc180-2697-11eb-928f-ccabac8c45bd.png">

mateuszmandera commented:
Looks great overall! I have just two points:
1. I think the two commits should be squashed - I don't see the value in splitting out a commit for adding fixtures, it actually seems more logical to add fixtures in the same one that changes what the webhook expects and adds tests that will use the fixtures.
2. Should we be assuming that all those keys are present in the payload? That'll generate exceptions if a key is not present. Perhaps we should use `.get(key)` to account for keys not being present and/or return `json_error` if we don't want to proceed without a key present.

maxrzaw commented:
Thanks for the feedback!
It is possible to modify the payload sent form NewRelic, so we should not assume all will be present, although it is unlikely that the fields I used would be removed as they are included by default. I changed it to use .get() with default values for most cases. I also changed the case for current_state to use json_error() instead of  the unsupported web hook event so the error message was more coherent.

mateuszmandera commented on zerver/webhooks/newrelic/view.py:
```diff
@@ -1,49 +1,61 @@
+        "incident_acknowledge_url": payload.get('incident_acknowledge_url', 'https://alerts.newrelic.com'),
+    }
+
+    status = payload.get('current_state', None)
+
+    if 'open' in status:
```
This will throw exceptions potentially `TypeError: argument of type 'NoneType' is not iterable`

We should check that status isn't None and return json_error early.

mateuszmandera replied:
It's a bit disappointing that mypy isn't detecting this :disappointed:

maxrzaw replied:
Yeah, that is disappointing. Added a test with that field missing.

mateuszmandera commented:
Looks good to me aside of the one error-handling comment. Though I don't work much with our webhook system, so perhaps other reviewers will have more comments.

timabbott commented:
Thanks for working on this!  How did you confirm that the old format is no longer supported?

I can confirm from our server logs that as of October 29, the existing webhook integration worked with at least one organization.

I'll also post a few comments on formatting.

timabbott commented on zerver/webhooks/newrelic/tests.py:
```diff
@@ -4,40 +4,123 @@
-        expected_topic = "Apdex score fell below critical level of 0.90"
-        expected_message = 'Alert opened on [application name]: Apdex score fell below critical level of 0.90 ([view alert](https://rpm.newrelc.com/accounts/[account_id]/applications/[application_id]/incidents/[incident_id])).'
+    def test_open(self) -> None:
+        expected_topic = "New Relic Alert - Test Policy (1234)"
```
Just to be clear, the "New Relic Alert" - Test Policy" is a name you entered in configuring New Relic, and doesn't have to contain "New Relic", right?  Generally or policy is to not repeat the name of a product here because it's usually implied by the name of the bot sending the message.

maxrzaw replied:
Correct, I will change the tests to reflect this policy and be more clear.

timabbott commented on zerver/webhooks/newrelic/tests.py:
````diff
@@ -4,40 +4,123 @@
+``` quote
+I am a test Violation
+```
+[View incident](https://alerts.newrelic.com/accounts/2941966/incidents/1234)
````
Can we just make the link on the word `Incident` at the start of the message?  That'd be more compact without any loss of information.

neiljp replied:
Do you think this would benefit from a topic link too?

timabbott replied:
Generally we avoid putting links in the topic as they make them very long.

timabbott commented on zerver/webhooks/newrelic/view.py:
```diff
@@ -1,49 +1,78 @@
+    except ValueError:
+        return json_error(_("The newrelic webhook expects timestamp and duration in milliseconds"))
+    except TypeError:
+        return json_error(_("The newrelic webhook expects timestamp and duration in milliseconds"))
```
Can we move the time-formatting logic for generating `<time` syntax to common code in `zerver/lib/webhooks/common.py`?  We'd need to make the exception use the name of the integration and JsonableError, but that'd make it easier to reuse for other webhooks.

maxrzaw replied:
I am not quite sure I understand this one. 
Would I add a subclass of JsonableError and then a unix_milleseconds_to_timestamp method, or would I just add the unix_milleseconds_to_timestamp method that takes in milliseconds and integration_name and throw a JsonableError?

timabbott replied:
I merged without addressing this since it's a fine follow-up PR.  The latter is the API I had in mind, and I think that's cleaner; there's no need for a special subclass that'll only be used in one place.

timabbott commented on zerver/webhooks/newrelic/tests.py:
```diff
@@ -4,40 +4,123 @@
+        expected_topic = "New Relic Alert - Test Policy (1234)"
         expected_message = """
-**1242** deployed by **Zulip Test**:
+Incident **closed** for condition: **Server Down**
```
Do we have the ability to linkify `Incident` here, or is the data not available?

maxrzaw replied:
Data is available!

timabbott commented on zerver/webhooks/newrelic/tests.py:
```diff
@@ -4,40 +4,123 @@
+    def test_acknowledged(self) -> None:
+        expected_topic = "New Relic Alert - Test Policy (1234)"
+        expected_message = """
+Incident **acknowledged** for condition: **Server Down**
```
Same question on linkification here.

maxrzaw commented:
> Thanks for working on this! How did you confirm that the old format is no longer supported?
> 
> I can confirm from our server logs that as of October 29, the existing webhook integration worked with at least one organization.
> 
> I'll also post a few comments on formatting.

This is the forum post where we were told that the old format is no longer in use.
https://discuss.newrelic.com/t/legacy-new-relic-web-hooks/120672
I am kinda surprised that it was used that recently since they didn't seem to know what the old format was. I wonder if someone was using a fixed url so they could get a message in Zulip. I can think of a way to hack the current system into working by using a different notification channel in New Relic for each alert.

devurandom commented:
> > Thanks for working on this! How did you confirm that the old format is no longer supported?
> >
> > I can confirm from our server logs that as of October 29, the existing webhook integration worked with at least one organization.
>
> This is the forum post where we were told that the old format is no longer in use.
https://discuss.newrelic.com/t/legacy-new-relic-web-hooks/120672
> I am kinda surprised that it was used that recently since they didn't seem to know what the old format was. I wonder if someone was using a fixed url so they could get a message in Zulip. I can think of a way to hack the current system into working by using a different notification channel in New Relic for each alert.

I asked them again to clarify, in case we misunderstood what they mean by "not supported": https://discuss.newrelic.com/t/legacy-new-relic-web-hooks/120672/8

timabbott commented:
It's possible that the one HTTP post we got was from someone doing something hacky (i.e. their own code is posting in the New Relic format), so given that clarity from the New Relic maintainers and the fact it was weeks ago, I'm up for merging this change as described.

timabbott commented:
Merged, thanks @maxrzaw!

Oh, can you also do a PR updating the documentation?  Note that https://github.com/zulip/zulip/pull/16682 exists, but it was done before we'd realized the webhook needed to change, and I'd guess needs some work (e.g. new screenshots generated by the tool, at least).

@punchagan FYI in case you need help with making the screenshots (I think we documented the automated process, but am not 100% sure).

punchagan commented:
> @punchagan FYI in case you need help with making the screenshots (I think we documented the automated process, but am 
not 100% sure).

@maxrzaw The last step in this section describes how to use the tool to generate screenshots. Feel free to ask me questions (here or on Zulip if you are stuck) or to improve the documentation of the tool.

https://zulip.readthedocs.io/en/latest/documentation/integrations.html#documenting-an-integration

devurandom commented:
Is this live already? I tried sending a test notification from New Relic, but it says the "Webhook responded with HTTP status code 400."

timabbott commented:
It's live on chat.zulip.org, but no on zulipchat.com yet (likely will be by Monday).

devurandom commented:
Thanks! Are changes to zulipchat.com announced somewhere, so I can get notified and test?
